File: netDxf/Entities/Helix.Geometry.cs
#region netDxf library licensed under the MIT License
// 
//                       netDxf library
// Copyright (c) Daniel Carvajal (haplokuon@gmail.com)
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
// 
#endregion

using System;
using System.Collections.Generic;
using netDxf.Tables;

namespace netDxf.Entities
{
    public sealed partial class Helix
    {
        /// <summary>Constructs a cylindrical/tapered helix or planar spiral with an error-bounded cubic Hermite spline.</summary>
        /// <param name="axisBasePoint">World-coordinate center at the start of the axis.</param>
        /// <param name="startPoint">World-coordinate start, in the plane perpendicular to the axis through its base.</param>
        /// <param name="axisVector">Finite nonzero world-coordinate axis; its original magnitude is retained.</param>
        /// <param name="radius">Nonnegative terminal radius. Initial radius comes from the start point.</param>
        /// <param name="turns">Positive number of turns, including fractional turns.</param>
        /// <param name="turnHeight">Signed height per turn; zero creates a planar spiral.</param>
        /// <param name="isRightHanded">True for right-handed angular progression.</param>
        /// <param name="tolerance">Positive absolute world-distance bound on the cubic approximation, excluding floating-point error.</param>
        /// <param name="maximumSegments">Allocation/work bound, from 1 through 1,048,576 cubic segments.</param>
        /// <returns>A new unowned HELIX whose stored spline approximates its analytic definition.</returns>
        public static Helix Create(Vector3 axisBasePoint, Vector3 startPoint, Vector3 axisVector,
            double radius, double turns, double turnHeight, bool isRightHanded = true,
            double tolerance = 1.0e-5, int maximumSegments = 65536)
        {
            return CreateCore(axisBasePoint, startPoint, axisVector, radius, turns, turnHeight,
                isRightHanded, tolerance, maximumSegments, null);
        }

        private static Helix CreateCore(Vector3 axisBasePoint, Vector3 startPoint, Vector3 axisVector,
            double radius, double turns, double turnHeight, bool isRightHanded,
            double tolerance, int maximumSegments, Vector3? phaseHint)
        {
            ValidateFinite(radius, nameof(radius)); ValidateFinite(turns, nameof(turns)); ValidateFinite(turnHeight, nameof(turnHeight));
            if (radius < 0.0) throw new ArgumentOutOfRangeException(nameof(radius));
            if (turns <= 0.0) throw new ArgumentOutOfRangeException(nameof(turns));
            DefinitionFrame frame = GetDefinitionFrame(axisBasePoint, startPoint, axisVector);
            if (frame.Radius == 0.0 && radius != 0.0 && phaseHint.HasValue) ApplyPhaseHint(ref frame, phaseHint.Value);
            double angle = (isRightHanded ? 2.0 : -2.0) * Math.PI * turns;
            double height = turns * turnHeight;
            ValidateFinite(angle, nameof(turns)); ValidateFinite(height, nameof(turnHeight));
            int segments = SelectSegments(frame.Radius, radius, turns, tolerance, maximumSegments);
            var curves = new List<BezierCurveCubic>(segments);
            Vector3 p0, d0; Evaluate(frame, axisBasePoint, radius, angle, height, 0.0, out p0, out d0);
            Vector3 firstDerivative = d0;
            for (int i = 0; i < segments; ++i)
            {
                Vector3 p1, d1;
                Evaluate(frame, axisBasePoint, radius, angle, height, (i + 1.0) / segments, out p1, out d1);
                Vector3 c1 = p0 + d0 / (3.0 * segments), c2 = p1 - d1 / (3.0 * segments);
                ValidateFinite(c1, nameof(tolerance)); ValidateFinite(c2, nameof(tolerance));
                curves.Add(new BezierCurveCubic(p0, c1, c2, p1));
                p0 = p1; d0 = d1;
            }
            var spline = new Spline(curves) { StartTangent = firstDerivative, EndTangent = d0 };
            return new Helix(spline)
            {
                AxisBasePoint = axisBasePoint, StartPoint = startPoint, AxisVector = axisVector,
                Radius = radius, Turns = turns, TurnHeight = turnHeight, IsRightHanded = isRightHanded,
                Normal = frame.Axis
            };
        }

        /// <summary>Evaluates the analytic parameter definition at a normalized parameter in [0,1], not the stored spline.</summary>
        public Vector3 EvaluateDefinition(double parameter)
        {
            Vector3 point, derivative;
            this.EvaluateDefinition(parameter, out point, out derivative);
            return point;
        }

        /// <summary>Evaluates the analytic derivative with respect to the normalized parameter; the vector is not normalized.</summary>
        public Vector3 EvaluateDefinitionDerivative(double parameter)
        {
            Vector3 point, derivative;
            this.EvaluateDefinition(parameter, out point, out derivative);
            return derivative;
        }

        private void EvaluateDefinition(double parameter, out Vector3 point, out Vector3 derivative)
        {
            ValidateFinite(parameter, nameof(parameter));
            if (parameter < 0.0 || parameter > 1.0) throw new ArgumentOutOfRangeException(nameof(parameter));
            DefinitionFrame frame = GetDefinitionFrame(this.axisBasePoint, this.startPoint, this.axisVector);
            if (frame.Radius == 0.0 && this.radius != 0.0)
            {
                Vector3? hint = this.GetDefinitionPhaseHint();
                if (hint.HasValue) ApplyPhaseHint(ref frame, hint.Value);
            }
            double angle = (this.IsRightHanded ? 2.0 : -2.0) * Math.PI * this.turns, height = this.turns * this.turnHeight;
            ValidateFinite(angle, nameof(this.Turns)); ValidateFinite(height, nameof(this.TurnHeight));
            Evaluate(frame, this.axisBasePoint, this.radius, angle, height, parameter, out point, out derivative);
        }

        /// <summary>Returns a conservative exact-arithmetic cubic Hermite deviation bound for equal parameter intervals.</summary>
        /// <remarks>This bounds a newly generated approximation with that segment count, not an arbitrary imported/edited spline.</remarks>
        public double GetApproximationErrorBound(int segments)
        {
            if (segments <= 0) throw new ArgumentOutOfRangeException(nameof(segments));
            DefinitionFrame frame = GetDefinitionFrame(this.axisBasePoint, this.startPoint, this.axisVector);
            return ErrorBound(frame.Radius, this.radius, this.turns, segments);
        }

        /// <summary>Returns a new helix with a regenerated error-bounded spline, retaining parameters, common metadata and XData.</summary>
        /// <remarks>The source, its database identity and its old spline arrays are not mutated. Old fit metadata is replaced, not merged.</remarks>
        public Helix WithRegeneratedSpline(double tolerance = 1.0e-5, int maximumSegments = 65536)
        {
            Helix result = CreateCore(this.axisBasePoint, this.startPoint, this.axisVector, this.radius, this.turns,
                this.turnHeight, this.IsRightHanded, tolerance, maximumSegments, this.GetDefinitionPhaseHint());
            result.MajorReleaseNumber = this.majorReleaseNumber; result.MaintenanceReleaseNumber = this.maintenanceReleaseNumber;
            result.Constraint = this.constraint;
            result.Layer = (Layer) this.Layer.Clone(); result.Linetype = (Linetype) this.Linetype.Clone();
            result.Color = (AciColor) this.Color.Clone(); result.Transparency = (Transparency) this.Transparency.Clone();
            result.Lineweight = this.Lineweight; result.LinetypeScale = this.LinetypeScale;
            result.IsVisible = this.IsVisible; result.Normal = this.Normal;
            result.KnotTolerance = this.KnotTolerance; result.CtrlPointTolerance = this.CtrlPointTolerance; result.FitTolerance = this.FitTolerance;
            foreach (XData data in this.XData.Values) result.XData.Add((XData) data.Clone());
            return result;
        }

        private struct DefinitionFrame
        {
            internal Vector3 Axis, Radial, Perpendicular;
            internal double Radius;
        }

        // At a zero-radius start the parameter subclass does not encode
        // radial phase. An authored spline tangent supplies that missing phase;
        // without a usable hint the deterministic arbitrary-axis frame is used.
        private Vector3? GetDefinitionPhaseHint()
        {
            return this.StartTangent;
        }

        private static Vector3 DivideComponents(Vector3 value, double scale)
        {
            return new Vector3(value.X / scale, value.Y / scale, value.Z / scale);
        }

        private static void ApplyPhaseHint(ref DefinitionFrame frame, Vector3 hint)
        {
            ValidateFinite(hint, nameof(hint));
            Vector3 radial = hint - frame.Axis * Vector3.DotProduct(hint, frame.Axis);
            double length = StableLength(radial);
            if (!IsFinite(length)) throw new ArgumentOutOfRangeException(nameof(hint));
            if (length == 0.0) return;
            frame.Radial = DivideComponents(radial, length);
            frame.Perpendicular = Vector3.CrossProduct(frame.Axis, frame.Radial);
        }

        private static DefinitionFrame GetDefinitionFrame(Vector3 axisBase, Vector3 start, Vector3 axis)
        {
            ValidateFinite(axisBase, nameof(axisBase)); ValidateFinite(start, nameof(start)); ValidateFinite(axis, nameof(axis));
            double length = StableLength(axis);
            if (!IsFinite(length) || length == 0.0) throw new ArgumentOutOfRangeException(nameof(axis));
            Vector3 n = DivideComponents(axis, length), radial = start - axisBase;
            ValidateFinite(radial, nameof(start));
            double r = StableLength(radial), axial = Vector3.DotProduct(radial, n);
            if (!IsFinite(r) || Math.Abs(axial) > 1.0e-10 * r)
                throw new ArgumentException("HELIX start point must lie in the plane through the axis base perpendicular to its axis.", nameof(start));
            radial -= n * axial; // Remove accepted floating-point orthogonality residual.
            r = StableLength(radial);
            Vector3 u = r == 0.0 ? MathHelper.ArbitraryAxis(n) * Vector3.UnitX : DivideComponents(radial, r);
            return new DefinitionFrame { Axis = n, Radial = u, Perpendicular = Vector3.CrossProduct(n, u), Radius = r };
        }

        private static void Evaluate(DefinitionFrame frame, Vector3 axisBase, double endRadius, double angle,
            double height, double parameter, out Vector3 point, out Vector3 derivative)
        {
            double deltaRadius = endRadius - frame.Radius;
            double radiusAt = frame.Radius * (1.0 - parameter) + endRadius * parameter;
            double phase = angle * parameter, cos = Math.Cos(phase), sin = Math.Sin(phase);
            Vector3 radial = frame.Radial * cos + frame.Perpendicular * sin;
            Vector3 angular = -frame.Radial * sin + frame.Perpendicular * cos;
            point = axisBase + frame.Axis * (height * parameter) + radial * radiusAt;
            derivative = frame.Axis * height + radial * deltaRadius + angular * (radiusAt * angle);
            ValidateFinite(point, nameof(parameter)); ValidateFinite(derivative, nameof(parameter));
        }

        private static double ErrorBound(double startRadius, double endRadius, double turns, int segments)
        {
            if (startRadius == 0.0 && endRadius == 0.0) return 0.0;
            // Evaluate the fourth-derivative/Hermite bound in log space so
            // large radius and tiny angular step cannot produce infinity*zero.
            double logStep = Math.Log(2.0 * Math.PI) + Math.Log(turns) - Math.Log(segments);
            double radial = Math.Log(Math.Max(startRadius, endRadius)) + 4.0 * logStep;
            double delta = Math.Abs(endRadius - startRadius);
            double taper = delta == 0.0 ? double.NegativeInfinity :
                Math.Log(4.0) + Math.Log(delta) - Math.Log(segments) + 3.0 * logStep;
            double max = Math.Max(radial, taper);
            double sum = max + Math.Log(Math.Exp(radial - max) + Math.Exp(taper - max));
            // The scalar Hermite remainder is h^4/384. sqrt(3) converts
            // component-wise fourth-derivative bounds to Euclidean distance.
            double bound = Math.Exp(sum + Math.Log(Math.Sqrt(3.0) / 384.0)) * (1.0 + 1.0e-12);
            if (double.IsPositiveInfinity(bound)) return bound;
            return BitConverter.Int64BitsToDouble(BitConverter.DoubleToInt64Bits(bound) + 1);

        }

        private static int SelectSegments(double startRadius, double endRadius, double turns, double tolerance, int maximumSegments)
        {
            ValidateFinite(tolerance, nameof(tolerance));
            if (tolerance <= 0.0) throw new ArgumentOutOfRangeException(nameof(tolerance));
            if (maximumSegments < 1 || maximumSegments > 1048576) throw new ArgumentOutOfRangeException(nameof(maximumSegments));
            double minimum = startRadius == 0.0 && endRadius == 0.0 ? 1.0 : Math.Max(1.0, Math.Ceiling(4.0 * turns));
            if (minimum > maximumSegments) throw new ArgumentException("HELIX turn count exceeds the supplied cubic segment budget.", nameof(maximumSegments));
            int segments = (int) minimum;
            while (ErrorBound(startRadius, endRadius, turns, segments) > tolerance)
            {
                if (segments == maximumSegments)
                    throw new ArgumentException("HELIX approximation cannot meet tolerance within the supplied segment budget.", nameof(maximumSegments));
                segments = Math.Min(maximumSegments, checked(segments * 2));
            }
            return segments;
        }
    }
}

File: javascript/netDxf/Entities/Helix.Geometry.js
// Copyright (c) Daniel Carvajal and netDxf contributors. MIT License; see package LICENSE.
import { Spline } from './Spline.js';
import { Vector3 } from '../Vector3.js';
import { Matrix3 } from '../Matrix3.js';
import { MathHelper } from '../MathHelper.js';
import { BezierCurveCubic } from '../BezierCurveCubic.js';
import { DotNetMath as M, MultiplyDouble as mul } from '../../runtime/GeometryRuntime.js';
import { ArgumentException, ArgumentOutOfRangeException, RequireInteger } from '../../runtime/Errors.js';
const bits=new DataView(new ArrayBuffer(8));
const divide=(v,s)=>new Vector3(v.X/s,v.Y/s,v.Z/s);
export function InstallHelixGeometry(Helix) {
  const valid=(v,p)=>Helix.ValidateFinite(v,p),length=v=>Helix.StableLength(v);
  function frame(axisBase,start,axis) {
    valid(axisBase,'axisBase');valid(start,'start');valid(axis,'axis');
    const magnitude=length(axis);if(!Number.isFinite(magnitude)||magnitude===0)throw new ArgumentOutOfRangeException('axis');
    const n=divide(axis,magnitude);let radial=Vector3.Subtract(start,axisBase);valid(radial,'start');
    let r=length(radial);const axial=Vector3.DotProduct(radial,n);
    if(!Number.isFinite(r)||Math.abs(axial)>mul(1e-10,r))throw new ArgumentException('HELIX start must lie in the perpendicular plane through the axis base.','start');
    radial=Vector3.Subtract(radial,Vector3.Multiply(n,axial));r=length(radial);
    const u=r===0?Matrix3.Multiply(MathHelper.ArbitraryAxis(n),Vector3.UnitX):divide(radial,r);
    return {Axis:n,Radial:u,Perpendicular:Vector3.CrossProduct(n,u),Radius:r};
  }
  function phase(f,hint) {
    valid(hint,'hint');const r=Vector3.Subtract(hint,Vector3.Multiply(f.Axis,Vector3.DotProduct(hint,f.Axis))),l=length(r);
    if(!Number.isFinite(l))throw new ArgumentOutOfRangeException('hint');if(l===0)return;
    f.Radial=divide(r,l);f.Perpendicular=Vector3.CrossProduct(f.Axis,f.Radial);
  }
  function evaluate(f,base,endRadius,angle,height,t) {
    const delta=endRadius-f.Radius,radius=mul(f.Radius,1-t)+mul(endRadius,t),a=mul(angle,t),cos=M.Cos(a),sin=M.Sin(a);
    const radial=Vector3.Add(Vector3.Multiply(f.Radial,cos),Vector3.Multiply(f.Perpendicular,sin));
    const angular=Vector3.Add(Vector3.Multiply(Vector3.Negate(f.Radial),sin),Vector3.Multiply(f.Perpendicular,cos));
    const point=Vector3.Add(Vector3.Add(base,Vector3.Multiply(f.Axis,mul(height,t))),Vector3.Multiply(radial,radius));
    const derivative=Vector3.Add(Vector3.Add(Vector3.Multiply(f.Axis,height),Vector3.Multiply(radial,delta)),Vector3.Multiply(angular,mul(radius,angle)));
    valid(point,'parameter');valid(derivative,'parameter');return {point,derivative};
  }
  function errorBound(startRadius,endRadius,turns,segments) {
    if(startRadius===0&&endRadius===0)return 0;
    const step=M.Log(mul(2,Math.PI))+M.Log(turns)-M.Log(segments),radial=M.Log(Math.max(startRadius,endRadius))+mul(4,step);
    const delta=Math.abs(endRadius-startRadius),taper=delta===0?-Infinity:M.Log(4)+M.Log(delta)-M.Log(segments)+mul(3,step),max=Math.max(radial,taper);
    const sum=max+M.Log(M.Exp(radial-max)+M.Exp(taper-max));
    const bound=mul(M.Exp(sum+M.Log(Math.sqrt(3)/384)),1+1e-12);if(bound===Infinity)return bound;
    bits.setFloat64(0,bound);bits.setBigUint64(0,bits.getBigUint64(0)+1n);return bits.getFloat64(0);
  }
  function segments(startRadius,endRadius,turns,tolerance,maximumSegments) {
    valid(tolerance,'tolerance');if(tolerance<=0)throw new ArgumentOutOfRangeException('tolerance');RequireInteger(maximumSegments,1,1048576,'maximumSegments');
    const minimum=startRadius===0&&endRadius===0?1:Math.max(1,Math.ceil(mul(4,turns)));
    if(minimum>maximumSegments)throw new ArgumentException('HELIX turns exceed the cubic segment budget.','maximumSegments');
    let n=minimum;while(errorBound(startRadius,endRadius,turns,n)>tolerance){
      if(n===maximumSegments)throw new ArgumentException('HELIX tolerance cannot be met within the cubic segment budget.','maximumSegments');
      n=Math.min(maximumSegments,n*2);
    }return n;
  }
  function create(base,start,axis,radius,turns,turnHeight,right,tolerance,budget,hint) {
    valid(radius,'radius');valid(turns,'turns');valid(turnHeight,'turnHeight');
    if(radius<0)throw new ArgumentOutOfRangeException('radius');if(turns<=0)throw new ArgumentOutOfRangeException('turns');
    const f=frame(base,start,axis);if(f.Radius===0&&radius!==0&&hint!==null)phase(f,hint);
    const angle=mul(mul(right?2:-2,Math.PI),turns),height=mul(turns,turnHeight);valid(angle,'turns');valid(height,'turnHeight');
    const count=segments(f.Radius,radius,turns,tolerance,budget),curves=[];
    let previous=evaluate(f,base,radius,angle,height,0);const first=previous.derivative;
    for(let i=0;i<count;i++){
      const next=evaluate(f,base,radius,angle,height,(i+1)/count);
      const c1=Vector3.Add(previous.point,Vector3.Divide(previous.derivative,mul(3,count))),c2=Vector3.Subtract(next.point,Vector3.Divide(next.derivative,mul(3,count)));
      valid(c1,'tolerance');valid(c2,'tolerance');curves.push(new BezierCurveCubic(previous.point,c1,c2,next.point));previous=next;
    }
    const spline=new Spline(curves);spline.StartTangent=first;spline.EndTangent=previous.derivative;
    const result=new Helix(spline);Object.assign(result,{AxisBasePoint:base,StartPoint:start,AxisVector:axis,Radius:radius,Turns:turns,TurnHeight:turnHeight,IsRightHanded:right,Normal:f.Axis});return result;
  }
  function definition(h,t) {
    valid(t,'parameter');if(t<0||t>1)throw new ArgumentOutOfRangeException('parameter');
    const f=frame(h.AxisBasePoint,h.StartPoint,h.AxisVector);if(f.Radius===0&&h.Radius!==0&&h.StartTangent!==null)phase(f,h.StartTangent);
    const angle=mul(mul(h.IsRightHanded?2:-2,Math.PI),h.Turns),height=mul(h.Turns,h.TurnHeight);valid(angle,'Turns');valid(height,'TurnHeight');
    return evaluate(f,h.AxisBasePoint,h.Radius,angle,height,t);
  }
  Helix.Create=(base,start,axis,radius,turns,turnHeight,right=true,tolerance=1e-5,maximumSegments=65536)=>create(base,start,axis,radius,turns,turnHeight,right,tolerance,maximumSegments,null);
  Helix.prototype.EvaluateDefinition=function(t){return definition(this,t).point;};
  Helix.prototype.EvaluateDefinitionDerivative=function(t){return definition(this,t).derivative;};
  Helix.prototype.GetApproximationErrorBound=function(count){RequireInteger(count,1,2147483647,'segments');return errorBound(frame(this.AxisBasePoint,this.StartPoint,this.AxisVector).Radius,this.Radius,this.Turns,count);};
  Helix.prototype.WithRegeneratedSpline=function(tolerance=1e-5,maximumSegments=65536){
    const r=create(this.AxisBasePoint,this.StartPoint,this.AxisVector,this.Radius,this.Turns,this.TurnHeight,this.IsRightHanded,tolerance,maximumSegments,this.StartTangent);
    for(const key of ['MajorReleaseNumber','MaintenanceReleaseNumber','Constraint'])r[key]=this[key];
    for(const key of ['Layer','Linetype','Color','Transparency'])r[key]=this[key].Clone();
    for(const key of ['Lineweight','LinetypeScale','IsVisible','Normal','KnotTolerance','CtrlPointTolerance','FitTolerance'])r[key]=this[key];
    for(const data of this.XData.Values)r.XData.Add(data.Clone());return r;
  };
}
